File: Week 6_ Python/Problems/sentimental-credit/credit.py
from cs50 import get_int

credit = get_int("Credit Card Number: ")
length = len(str(credit))

# Calculate sum of digits
sum = 0

# For every digit
for i in range(length):
    digit = credit // 10**i % 10
    if i % 2 == 0:
        sum += digit
    else:
        other = digit * 2
        other1s = other % 10
        other10s = other // 10
        sum += other1s + other10s

# Check if sum has last digit 0
if sum % 10 == 0:
    # Find first two digits
    d1 = credit // 10**(length - 1)
    d2 = credit // 10**(length - 2) % 10

    # VISA
    if (length == 13 or length == 16) and d1 == 4:
        print("VISA")
    # American Express
    elif length == 15 and d1 == 3 and (d2 == 4 or d2 == 7):
        print("AMEX")
    # Mastercard
    elif length == 16 and d1 == 5 and d2 >= 1 and d2 <= 5:
        print("MASTERCARD")
    # Invalid (Doesn't match requirements)
    else:
        print("INVALID")
# Invalid (Doesn't end with 0)
else:
    print("INVALID")

File: Week 1_ C/Problems/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main(void)
{
    // Declare variables
    long credit = get_long("Credit Card Number: ");
    int length = 0;
    long creditCount = credit;

    // Get Length of Number
    while (creditCount > 0)
    {
        // Divide by 10 until int < 0 while incrementing length
        creditCount = creditCount / 10;
        length++;
    }

    // Calculate sum of digits using Luhn's Algorithm
    int sum = 0;

    // For every digit
    for (int i = 0; i < length; i++)
    {
        // Find ith digit
        long power = pow(10, i);
        int digit = (credit / power) % 10;

        // If every other location
        if (i % 2 == 0)
        {
            sum += digit;
        }
        else
        {
            int otherDigit = digit * 2;
            int otherOnes = otherDigit % 10;
            int otherTens = otherDigit / 10;
            int otherSum = otherOnes + otherTens;
            sum += otherSum;
        }
    }

    // Rule: Sum has last digit 0
    if (sum % 10 == 0)
    {
        // Determine which company credit card belongs to
        long power = pow(10, length - 1);

        int d1 = (credit / power) % 10; // digit 1
        int d2 = (credit / (power / 10)) % 10; // digit 2

        // VISA
        if ((length == 13 || length == 16) && d1 == 4)
        {
            printf("VISA\n");
        }
        // American Express
        else if (length == 15 && d1 == 3 && (d2 == 4 || d2 == 7))
        {
            printf("AMEX\n");
        }
        // Mastercard
        else if (length == 16 && d1 == 5 && d2 >= 1 && d2 <= 5)
        {
            printf("MASTERCARD\n");
        }
        // Invalid (Doesn't match requirements)
        else
        {
            printf("INVALID\n");
        }
    }
    // Invalid (Doesn't end with 0)
    else
    {
        printf("INVALID\n");
    }
}